src/Dialect/ONNX/Transforms/Decompose.cpp

```diff
@@ -1581,6 +1582,72 @@ struct CustomOpFuseMatMulPattern : public OpRewritePattern<ONNXCustomOp> {
   }
 };
 
+namespace {
+
+[[nodiscard]] bool isCustomMicrosoftOp(
+    ONNXCustomOp customOp, StringRef expectedName) {
+  if (!customOp.getFunctionName().equals_insensitive(expectedName)) {
+    return false;
+  }
+
+  const auto domAttr = customOp->getAttrOfType<StringAttr>("domain_name");
+  return domAttr && domAttr.getValue().equals_insensitive("com.microsoft");
+}
+
+} // namespace
+
+template <typename OpToCreate, typename Derived>
+struct CustomOpMicrosoftQDuantizeLinear
+    : public OpRewritePattern<ONNXCustomOp> {
+  using OpRewritePattern<ONNXCustomOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      ONNXCustomOp customOp, PatternRewriter &rewriter) const final {
+    using namespace onnx_mlir;
+
+    if (!isCustomMicrosoftOp(
+            customOp, static_cast<const Derived *>(this)->expectedName))
+      return failure();
+    assert(customOp->getNumOperands() == 3);
+
+    const auto scale = customOp->getOperand(1);
+    const auto zeroPoint = customOp->getOperand(2);
+    if (!isScalarTensor(scale) || !isScalarTensor(zeroPoint)) {
+      return rewriter.notifyMatchFailure(
+          customOp, "Only supports per-tensor quantization for now");
+    }
+    // Axis is ignored if scale and zeroPoint are scalars
+
+    auto newOp = rewriter.create<OpToCreate>(customOp->getLoc(),
+        customOp.getResult(0).getType(), customOp->getOperand(0), scale,
+        zeroPoint);
+
+    IgnoreDiagnostic diag(customOp->getContext()->getDiagEngine());
+    if (failed(mlir::verify(newOp))) {
```

Comment:
I am using something very similar to legalize QuarkQuantized ops, and I noticed that this doesn't trigger all checkers we can have for the operation. For example, if the op require inputs to have the same type. To get op-specific verification you can use something like this:
```suggestion
    bool isNewOpValid;
    if (auto info = newOp->getName().getRegisteredInfo()) {
      isNewOpValid= succeeded(info->verifyInvariants(newOp));
    } else {
      isNewOpValid= succeeded(mlir::verify(newOp));
    }
```

Reply:
Why are we manually verifying ops? I believe MLIR already does it quite often, what am I missing?

Reply:
we only want to to do the rewrite if we can successfully create the new op. This is not always possible, as they support different types. Instead of hardcoding a list of supported types, we can use the verifier to check it for us